File: MONTH-01 FOLDER/DAY-12/palindrome_solver.py
# Day 12: Valid Palindrome Algorithm using Python Two-Pointer Technique
# Time Complexity: O(n) | Space Complexity: O(1)

def is_palindrome(s: str) -> bool:
    # 1. Sanitize the string using list comprehension (Keep only alphanumeric lowercase)
    clean_chars = [char.lower() for char in s if char.isalnum()]
    
    # 2. Two-Pointer Execution Matrix
    left, right = 0, len(clean_chars) - 1
    
    while left < right:
        if clean_chars[left] != clean_chars[right]:
            return False  # Logical mismatch detected
        left += 1
        right -= 1
        
    return True

# Test cases evaluating the algorithm
test_string = "A man, a plan, a canal: Panama"
result = is_palindrome(test_string)
print(f"🐍 Python Algorithm Result for '{test_string}': {result}")

File: MONTH-01 FOLDER/DAY-12/palindromeSolver.js
// Day 12: Valid Palindrome Algorithm using JavaScript Native Array Mechanics
// Time Complexity: O(n) | Space Complexity: O(n)

function isPalindrome(s) {
    // 1. Sanitize string using regular expressions (Regex) and convert to lowercase
    const cleanStr = s.toLowerCase().replace(/[^a-z0-9]/g, "");
    
    // 2. Reverse comparison execution loop
    const reversedStr = cleanStr.split("").reverse().join("");
    
    return cleanStr === reversedStr;
}

// Test cases evaluating the runtime loop
const testString = "A man, a plan, a canal: Panama";
const result = isPalindrome(testString);
console.log(`⚡ JS Engine Verification for '${testString}': ${result}`);